Add YaRN and Dynamic-YaRN RoPE Scaling Methods

mig-mfreitas opened the pull request:
# What does this PR do?

YaRN (Yet another RoPE extension method) combines the NTK-By-Parts Interpolation and Attention Scaling methods, improving upon existing RoPE interpolation methods for longer context window sizes.

Fine-tuned models maintain their original performance across benchmarks while enabling efficient extrapolation and transfer learning for quicker convergence, especially in compute-limited environments.

We implement YaRN and Dynamic-YaRN for the following list of models:

 - LLaMA
 - Falcon
 - GPT-NeoX
 - Olmo
 - Persimmon
 - Phi
 - StableLM
 - OpenLLaMA

New unit tests are added to assert YaRN's correct behavior on both short and long sequence inputs.

For more details, please refer to https://arxiv.org/abs/2309.00071.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?
@gante

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

mig-mfreitas committed:
Add YaRN and Dynamic-YaRN RoPE Scaling Methods

YaRN (Yet another RoPE extension method) combines the NTK-By-Parts
Interpolation and Attention Scaling methods, improving upon existing
RoPE interpolation methods for longer context window sizes.

Fine-tuned models maintain their original performance across benchmarks
while enabling efficient extrapolation and transfer learning for
quicker convergence, especially in compute-limited environments.

We implement YaRN and Dynamic-YaRN for the following list of models:

 - LLaMA
 - Falcon
 - GPT-NeoX
 - Olmo
 - Persimmon
 - Phi
 - StableLM
 - OpenLLaMA

New unit tests are added to assert YaRN's correct behavior on both
short and long sequence inputs.

For more details, please refer to https://arxiv.org/abs/2309.00071.

Co-authored-by: Miguel Almeida <miguel.pessanha.almeida@tecnico.ulisboa.pt>

amyeroberts commented:
cc @ArthurZucker too

ArthurZucker commented:
Hey! Thanks a lot for taking the time to implement this! 🤗 
cc @gante with this and phi3's new method, not sure it makes sense to add all of them in the modeling code, WDYT? 

This is unrelated to Llama so we might need some modularity for this!

miguelm-almeida commented:
🤗

gante commented on src/transformers/models/falcon/configuration_falcon.py:
```diff
@@ -77,13 +77,31 @@ class FalconConfig(PretrainedConfig):
+                    Parameter to set the boundary for extrapolation (only) in the linear ramp function.
+                beta_slow (`float`, *optional*, defaults to 1):
+                    Parameter to set the boundary for interpolation (only) in the linear ramp function.
+                finetuned (`bool`, *optional*, defaults to `False`):
```
This can also be removed (see the comments on the new dynamic class)

gante commented on src/transformers/models/falcon/modeling_falcon.py:
```diff
@@ -209,6 +209,162 @@ def _set_cos_sin_cache(self, seq_len, device, dtype):
+            self._sin_cached[:seq_len, ...].to(dtype=x.dtype),
+        )
+
+    def yarn(self, device):
```
```suggestion
    def compute_yarn_scaling(self, device):
```
(Or a similar name. Let's use descriptive function names :) )

gante commented on src/transformers/models/falcon/modeling_falcon.py:
```diff
@@ -209,6 +209,162 @@ def _set_cos_sin_cache(self, seq_len, device, dtype):
+            device,
+        )
+
+        if finetuned:
```
```suggestion
        if self.max_position_embeddings != self.original_max_position_embeddings:
```
This should be true for fine-tuned models, saving us a flag :)

gante commented on tests/models/falcon/test_modeling_falcon.py:
```diff
@@ -510,6 +512,40 @@ def test_model_rope_scaling(self):
+        yarn_cos_long, yarn_sin_long = yarn_scaling_rope(x, long_input_length)
+        torch.testing.assert_close(yarn_cos_short, yarn_cos_long[:short_input_length, :])
+        torch.testing.assert_close(yarn_sin_short, yarn_sin_long[:short_input_length, :])
+        with self.assertRaises(AssertionError):
+            torch.testing.assert_close(yarn_cos_long, original_cos_long)
+        with self.assertRaises(AssertionError):
+            torch.testing.assert_close(yarn_sin_long, original_sin_long)
```
let's also check that yarn_sin/cos_short != original_sin/cos_short (i.e. that applying yarn should change all values)

gante reviewed (commented):
Olá Miguel e Miguel! 👋 (@miguelm-almeida @mig-mfreitas )

I have a couple of requests regarding user experience and recent changes in our repo. My number 1 suggestion would be to delete the diff in all models except `Llama`, and leave the model copies for another PR. It's much faster for everyone (you and me) to iterate over a model, and then copy the design when we're happy 🤗 In this particular case (RoPE models), we also have different implementations that we need to iron out on our end before adding Yarn there.

⚠️ please treat all my comment as if there were made on `Llama`, and not on `Falcon`. Some of the suggested changes only work on architectures that are up to date, like `Llama` (and unlike `Falcon`)

Finally: one of the goals of this PR should be to be able to load the original YaRN models using `transformers`. Currently, there are some models on the Hub that have custom code (e.g. https://huggingface.co/NousResearch/Yarn-Llama-2-7b-64k). At the moment, these models require adding `trust_remote_code=True` to `from_pretrained` (which loads the custom code in the repo). With this PR, we remove the need for that flag and would be using the code in `transformers` instead :)

Ping me if you have further questions (and feel free to ping me by email if I'm taking to long to reply) 🤗

miguelm-almeida commented:
Thank you very much for this in-depth review and suggestions! We'll iterate on it and reach back shortly 🤗

Commit:
Merge remote-tracking branch 'upstream/main' into yarn-rope-scaling

Commit:
Merge remote-tracking branch 'upstream/main' into yarn-rope-scaling

Commit:
Refactor YaRN implementation for LLaMA

Iterate on YaRN implementation for LLaMA and remove diff from remaining
models for increased PR modularity.

This commit includes the following changes:
- Merge 'yarn_rope_scaling' and 'rope_scaling' dictionaries
- Remove unnecessary attributes ('extrapolation_factor' and 'finetuned')
  from YaRN classes
- Inherit 'forward' method in YaRN classes from superclass
- Rename 'yarn' method to 'compute_yarn_scaling'
- Extend YaRN tests with further assertions
- Fix style inconsistencies

Co-authored-by: Miguel Monte e Freitas <miguelmontefreitas@tecnico.ulisboa.pt>

miguelm-almeida commented:
@gante Hi! We believe we've covered the mentioned topics in this iteration. Please tell us what you think! 🤗

gante commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -149,6 +149,140 @@ def forward(self, x, position_ids):
+        self.beta_fast = beta_fast
+        self.beta_slow = beta_slow
+
+        if self.attention_factor is None:
+            self.attention_factor = 0.1 * math.log(scaling_factor) + 1.0
```
Let's add a note saying that this is the default value according to the yarn paper, so it doesn't look like a magic number :)

gante commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -149,6 +149,140 @@ def forward(self, x, position_ids):
+        self.max_seq_len_cached = max_position_embeddings
+        emb = self.get_pos_embeddings(device)
+
+        self._cos_cached = (emb.cos() * self.mscale)[None, :, :].to(torch.get_default_dtype())
+        self._sin_cached = (emb.sin() * self.mscale)[None, :, :].to(torch.get_default_dtype())
```
The logic to build these tensors was removed in a recent PR (https://github.com/huggingface/transformers/pull/30743) -- we can delete them as well as all related functions/variables (`emb`, `get_pos_embeddings`, ...)

This comment applies to the other class too :)

gante reviewed (commented):
In general LGTM (aside two minor cleanup tasks) 👍 @mig-mfreitas @miguelm-almeida 

Before I give the green light, I'd like you to run a sanity check: does the original implementation of YaRN match this implementation? This check doesn't fit as a test (it probably takes too long), but I'd like you to share a script with us that shows that the two implementations generate the same output text :)

After we confirm that they are matching, we can add a slow integration test to ensure we don't regress.

ArthurZucker reviewed (commented):
Hey! Great work, I think here we have a philosophical choice:
we should start thinking about switching to sin and cos being computed once, at the beginning of the modeling, and that is is. 

ROPE scaling is mostly used during generation, thsu `generate` should be able to handle any rope scaling + we don't re-compute it every forward layer. 

If you want to work on that would be truly nice, and aligned with the FA2 refactor / our idea to add kwargs to the foward.

gante commented on src/transformers/models/llama/configuration_llama.py:
```diff
@@ -178,15 +187,52 @@ def _rope_scaling_validation(self):
             )
         if rope_scaling_factor is None or not isinstance(rope_scaling_factor, float) or rope_scaling_factor <= 1.0:
             raise ValueError(f"`rope_scaling`'s factor field must be a float > 1, got {rope_scaling_factor}")
+
+        if rope_scaling_type not in ["yarn", "dynamic-yarn"]:
+            return
+
+        if not isinstance(self.rope_scaling, dict) or len(self.rope_scaling) > 6:
+            raise ValueError(
+                "`rope_scaling` with type "
+                f"{rope_scaling_type}"
+                " must be a dictionary with a maximum of six fields, `type`, `factor`,"
+                "`original_max_position_embeddings`, `attention_factor`, `beta_fast`, `beta_slow`, "
+                f"got {self.rope_scaling}"
+            )
+        original_max_position_embeddings = self.rope_scaling.get("original_max_position_embeddings", None)
+        attention_factor = self.rope_scaling.get("attention_factor", None)
+        beta_fast = self.rope_scaling.get("beta_fast", None)
+        beta_slow = self.rope_scaling.get("beta_slow", None)
+
+        if original_max_position_embeddings is not None and not isinstance(original_max_position_embeddings, int):
+            raise ValueError(
+                f"`rope_scaling`'s original_max_position_embeddings field must be an int, got {original_max_position_embeddings}"
+            )
+        if attention_factor is not None and not isinstance(attention_factor, float) or attention_factor < 0:
+            raise ValueError(
+                f"`rope_scaling`'s attention_factor field must be a float greater than 0, got {attention_factor}"
+            )
+        if beta_fast is not None and not isinstance(beta_fast, float):
+            raise ValueError(f"`rope_scaling`'s beta_fast field must be a float, got {beta_fast}")
+        if beta_slow is not None and not isinstance(beta_slow, float):
+            raise ValueError(f"`rope_scaling`'s beta_slow field must be a float, got {beta_slow}")
+
+        b_fast = beta_fast if beta_fast is not None else 32
+        b_slow = beta_slow if beta_slow is not None else 1
+        if b_fast < b_slow:
+            raise ValueError(
+                f"`rope_scaling`'s beta_fast field must be greater than beta_slow, got beta_fast={b_fast} and beta_slow={b_slow}"
+            )
```
Agreed (but let's do it in a separate PR, to avoid adding library reorganization tasks on top of a 2 month old contributor PR :P )

ArthurZucker replied:
It's pretty much against our philosophy

Commit:
Refactor Tensor Building Logic for YaRN

- Comply with the the tensor building logic introduced in #30743
- Add referencing to the optimized Attention Factor equation
- Remove Dynamic YaRN for a more agile deployment

Co-authored-by: mig-mfreitas <mig-mfreitas@users.noreply.github.com>

gante commented:
Chatted with @ArthurZucker offline: let's prepare to merge this PR as is, after removing `dynamic-yarn` (which does not match the original implementation). cc @mig-mfreitas 

In parallel, I will be working on a PR on top of this one to move RoPE scaling code out of the models themselves, into a dedicated place. RoPE scaling would then behave as a plug in to models with RoPE. 

When the two PRs are ready: this PR will get merged and, straight away, mine is applied on top of it. We'll then have YaRN from @mig-mfreitas and the new structure from my PR :)

Commit:
Merge remote-tracking branch 'upstream/main' into yarn-rope-scaling

miguelm-almeida commented:
Done! 🤗🚀

gante commented:
@mig-mfreitas working on the new PR on top of this one today 🤗

gante committed:
remove unwanted file

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_30910). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

gante commented:
@mig-mfreitas @miguelm-almeida :D 

![image](https://github.com/user-attachments/assets/f7e4618c-758e-4fff-a985-3412d82dfbff)

miguelm-almeida commented:
😊😊😊

miguelm-almeida commented:
Thank you very much for all the support!

mig-mfreitas commented:
Thank you for all the guidance, this contribution has been a great experience 🤗🚀
